src/Build/Logging/BinaryLogger/BinaryLogReplayEventSource.cs

```diff
@@ -85,24 +198,136 @@ public void Replay(string sourceFilePath, CancellationToken cancellationToken)
+                reader.SkipUnknownEvents = supportsForwardCompatibility && AllowForwardCompatibility;
+                reader.SkipUnknownEventParts = supportsForwardCompatibility && AllowForwardCompatibility;
+                reader.RecoverableReadError += RecoverableReadError;
 
-            while (!cancellationToken.IsCancellationRequested && reader.Read() is { } instance)
+                while (!cancellationToken.IsCancellationRequested && reader.Read() is { } instance)
+                {
+                    Dispatch(instance);
+                }
+            }
+            else
             {
-                Dispatch(instance);
+                if (this._rawLogRecordReceived == null &&
+                    this._embeddedContentRead == null &&
+                    this._stringReadDone == null &&
+                    this._archiveFileEncountered == null)
+                {
+                    throw new NotSupportedException(
+                        ResourceUtilities.GetResourceString("Binlog_Source_MissingSubscribeError"));
+                }
+
+                while (!cancellationToken.IsCancellationRequested && reader.ReadRaw() is { } instance &&
+                       instance.RecordKind != BinaryLogRecordKind.EndOfFile)
+                {
+                    _rawLogRecordReceived?.Invoke(instance.RecordKind, instance.Stream);
+                }
             }
+
+            // Unsubscribe from events for a case if the reader is reused (upon cancellation).
+            reader.EmbeddedContentRead -= _embeddedContentRead;
+            reader.ArchiveFileEncountered -= _archiveFileEncountered;
+            reader.StringReadDone -= _stringReadDone;
+            reader.RecoverableReadError -= RecoverableReadError;
+        }
+
+        /// <inheritdoc cref="IRawLogEventsSource.DeferredInitialize"/>
+        void IRawLogEventsSource.DeferredInitialize(
+            Action onRawReadingPossible,
+            Action onStructuredReadingOnly)
+        {
+            this._onRawReadingPossible += onRawReadingPossible;
+            this._onStructuredReadingOnly += onStructuredReadingOnly;
+        }
+
+        public int FileFormatVersion => _fileFormatVersion ?? throw new InvalidOperationException(ResourceUtilities.GetResourceString("Binlog_Source_VersionUninitialized"));
+        public int MinimumReaderVersion => _minimumReaderVersion ?? throw new InvalidOperationException(ResourceUtilities.GetResourceString("Binlog_Source_VersionUninitialized"));
+
+        private int? _fileFormatVersion;
+        private int? _minimumReaderVersion;
+        private Action? _onRawReadingPossible;
+        private Action? _onStructuredReadingOnly;
+        private Action<EmbeddedContentEventArgs>? _embeddedContentRead;
+        /// <inheritdoc cref="IEmbeddedContentSource.EmbeddedContentRead"/>
+        event Action<EmbeddedContentEventArgs>? IEmbeddedContentSource.EmbeddedContentRead
```

Comment:
Why are these implemented explicitly?

Reply:
It's to avoid the need f exposing them publicly.

Totally unclear from that code - so I'm adding an explicit claryfying comment.